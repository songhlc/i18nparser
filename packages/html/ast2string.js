import * as html from 'html5parser'
function ast2String (ast, text) {
  if (!text) {
    text = ''
  }
  ast.forEach(item => {
    if (item.type === html.SyntaxKind.Text) {
      text += item.value
    } else if (item.type === html.SyntaxKind.Tag) {
      text += '<' + item.name
      item.attributes && item.attributes.forEach(attr => {
        if (attr.value)
          text += ' ' + attr.name.value + '=' + attr.value.quote + attr.value.value + attr.value.quote
      })
      if (item.name.toLowerCase() == '!doctype') {
        text += ' html'
      }
      text += '>'
      if (item.body) {
        text = ast2String(item.body, text)
      }
      if (item.close) {
        text += item.close.value
      }
    }
  })
  return text
}
export default ast2String
