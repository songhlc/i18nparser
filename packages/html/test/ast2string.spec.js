import ast2string from '../ast2string'
import * as html from 'html5parser-fork'

describe('测试ast2string', () => {
  // 这是mocha的玩法，jest可以直接兼容
  it('转换结果相等', () => {
    const input = ` <!doctype html>
    <html>
      <body>
        <h1 id="hello">Hello world</h1>
        <h2 title="有没有翻译">中文翻译</h2>
        <h3 data-bind="text: '你好'">世界</h3>
        <script>
          window.alert('error')
          function aaa (p) {
            var b = p + '1'
            console.log(b);
          } 
        </script>
      </body>
    </html>
    `;
    var ast = html.parse(input)
    var result = ast2string(ast)
    debugger
    // 这里的断言实际上和chai的expect是很像的
    expect(result == input).toBe(true);
    expect(result).toMatchSnapshot()
  })
})
