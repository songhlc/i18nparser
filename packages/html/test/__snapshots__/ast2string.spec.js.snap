// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试ast2string 转换结果相等 1`] = `
" <!doctype html>
    <html>
      <body>
        <h1 id=\\"hello\\">Hello world</h1>
        <h2 title=\\"有没有翻译\\">中文翻译</h2>
        <h3 data-bind=\\"text: '你好'\\">世界</h3>
        <script>
          window.alert('error')
          function aaa (p) {
            var b = p + '1'
            console.log(b);
          } 
        </script>
      </body>
    </html>
    "
`;
